include/swift/Sema/Constraint.h

```diff
@@ -158,6 +158,11 @@ enum class ConstraintKind : char {
   /// The first type is an optional type whose object type is the second
   /// type, preserving lvalue-ness.
   OptionalObject,
+  /// The first type is an optional of non-negative depth of the second type,
+  /// ignoring lvalueness (because the constraint is currently applied only to
+  /// pattern types, which are always rvalue). For example, given a RHS of
+  /// 'Int', a LHS of 'Int' or 'Int??' would satisfy this constraint.
+  EqualOrOptional,
```

Comment:
It doesn't quite feel right to me to have `Equal` in the name since the constraint isn't doing rvalue conversions, it's just asserting that no lvalues are present. Additionally, I think if we did ever want to apply this constraint to something else that could have lvalues, I think it would _probably_ make most sense to preserve lvalueness like `OptionalObject`, as lvalues are generally preserved through optional chains.

How about something like `UnwrappingOptionals`? I think the plurality also helps make it clear it unwraps an arbitrary number of optionals.